enhancements/ingress/global-admission-hsts.md

```diff
@@ -0,0 +1,631 @@
+
+## Proposal
+Each goal is addressed in the subsections below.
+
+### Continue to enable HSTS per-route (retain the existing feature functionality)
+The `Strict-Transport-Security` response header is currently applied per-route via an
+annotation called `haproxy.router.openshift.io/hsts_header`.  This annotation has a
+required directive `max-age`, and optional directives `includeSubDomains` and `preload`,
+with the following definitions:
+- `max-age`: delta time range in seconds during which the host is to be regarded as an HSTS host
+  - If set to 0, it negates the effect and the host is no longer regarded as an HSTS host
+  - `max-age` is a time-to-live value, so if a client makes a request for the route's host
+  and this period of time elapses before the client makes another request for the same host, the
+  HSTS policy will expire on that client
+- `includeSubDomains`: if present, applies to any subdomains of the host's domain name, when the
+route admission policy allows wildcards.  If wildcard routes aren't allowed, this directive has no
+effect.
+- `preload`: if present, tells the client to include the route's host in its host preload list so that
+it never needs to do an initial load to get the HSTS header (note that this is not defined in RFC 6797
+and is therefore client browser implementation-dependent)
+
+HSTS is currently implemented in the HAProxy template and applied to `edge` and
+`reencrypt` routes that have the `haproxy.router.openshift.io/hsts_header` annotation:
+```gotemplate
+{{- /* hsts header in response: */}}
+ {{- $hstsOptionalTokenPattern := `(?:includeSubDomains|preload)` }}
+ {{- $hstsPattern := printf `(?:%[1]s[;])*max-age=(?:\d+|"\d+")(?:[;]%[1]s)*`  $hstsOptionalTokenPattern -}}
+...
+{{- if matchValues (print $cfg.TLSTermination) "edge" "reencrypt" }}
+    {{- with $hsts := firstMatch $hstsPattern (index $cfg.Annotations "haproxy.router.openshift.io/hsts_header") }}
+  http-response set-header Strict-Transport-Security {{$hsts}}
+    {{- end }}{{/* hsts header */}}
+  {{- end }}{{/* is "edge" or "reencrypt" */}}
+```
+The HAProxy template will not change for this enhancement.  All routes must still be configured with
+the HSTS annotation.
+
+#### Provide HSTS verification per-domain
+This proposal allows cluster administrators to configure HSTS verification on a per-domain basis with
+the addition of a new openshift/api-server validating admission plugin for the router, called
+`route.openshift.io/RequiredRouteAnnotations`.  Each route must be configured with a required route
+annotation, which will be verified against the global setting on the cluster `Ingress` configuration.
+
+The administrator will interact with the `RequiredRouteAnnotations` plugin by configuring the
+`Ingress.Spec.RequiredHSTSPolicies` with one or more `RequiredHSTSPolicy`.  `RequiredHSTSPolicy` is
+a new type that will be added to the API `openshift/api/operator/v1/types_ingress.go`, to capture the
```

Comment:
Can we compare against https://github.com/openshift/api/blob/master/config/v1/types_ingress.go ?  By configuring per-domain, the configuration could be made fully global.

Reply:
Yes, this was a typo.